autoprecompiles/src/trace_handler.rs

```diff
@@ -0,0 +1,111 @@
+use itertools::Itertools;
+use rayon::prelude::*;
+use std::cmp::Eq;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use crate::Apc;
+
+/// Returns data needed for constructing the APC trace.
+pub struct TraceHandlerData<'a, F> {
+    /// The dummy trace values for each instruction.
+    pub dummy_values: Vec<Vec<&'a [F]>>,
+    /// The mapping from dummy trace index to APC index for each instruction.
+    pub dummy_trace_index_to_apc_index_by_instruction: Vec<HashMap<usize, usize>>,
+}
+
+pub trait TraceHandler {
+    type AirId: Hash + Eq + Sync;
+    type Field: Sync + Clone + Ord + std::fmt::Display;
+    type Instruction;
+
+    /// Returns a vector with the same length as original instructions
+    fn original_instruction_air_ids(&self) -> Vec<Self::AirId>;
+
+    /// Returns the number of APC calls, which is also the number of rows in the APC trace
+    fn apc_call_count(&self) -> usize;
+
+    /// Returns a mapping from air_id to the dummy trace
+    fn air_id_to_dummy_trace_and_width(&self) -> &HashMap<Self::AirId, DummyTrace<Self::Field>>;
```

Comment:
This is still DummyTrace, should be Trace. Also the method name should be `air_id_to_dummy_trace`